### PR TITLE
add reverse dns transport

### PR DESCRIPTION
**Pull request by busterb**
I thought I'd go ahead and create a PR for review from the @max3raza dns transport fork, after having looked it over for a bit.

### CODE REVIEW

File: c/meterpreter/source/common/config.h [hunk 1]
```diff
@@ -43,6 +46,16 @@ typedef struct _MetsrvTransportCommon
 	int retry_wait;                       ///! Seconds to wait between reconnects.
 } MetsrvTransportCommon;
 
+//typedef struct _MetsrvTransportCommonDns
```
Comment (max3raza) [inline c/meterpreter/source/common/config.h]: This structure should be deleted as not used anymore.

File: c/meterpreter/source/common/core.h [hunk 2]
```diff
@@ -160,7 +160,6 @@ typedef enum
 	TLV_TYPE_TRANS_RETRY_WAIT    = TLV_VALUE(TLV_META_TYPE_UINT,      440),   ///! Time (seconds) to wait between reconnect attempts.
 	TLV_TYPE_TRANS_HEADERS       = TLV_VALUE(TLV_META_TYPE_STRING,    441),   ///! List of custom headers to send with the requests.
 	TLV_TYPE_TRANS_GROUP         = TLV_VALUE(TLV_META_TYPE_GROUP,     442),   ///! A single transport grouping.
-
```
Comment (max3raza) [inline c/meterpreter/source/common/core.h]: Need to revert this changes.

File: c/meterpreter/source/common/remote.h [hunk 3]
```diff
@@ -99,6 +103,19 @@ typedef struct _HttpTransportContext
 	PReadResponse read_response;          ///! WinHttp/WinINET specific response data reading.
 } HttpTransportContext;
 
+typedef struct _DnsTransportContext
+{
+	BOOL ready; 
+	WORD request_type;
+	wchar_t *domain;                         ///! Pointer to the DNS domain stored with the transport.
+	wchar_t *ns_server;
+    wchar_t *client_id;
```
Comment (max3raza) [inline c/meterpreter/source/common/remote.h]: Inconsistent indentation.

File: c/meterpreter/source/server/server_setup_win.c [hunk 4]
```diff
@@ -75,9 +76,12 @@ VOID load_stageless_extensions(Remote* remote, MetsrvExtension* stagelessExtensi
 	while (stagelessExtensions->size > 0)
 	{
 		dprintf("[SERVER] Extension located at 0x%p: %u bytes", stagelessExtensions->dll, stagelessExtensions->size);
-		HMODULE hLibrary = LoadLibraryR(stagelessExtensions->dll, stagelessExtensions->size);
-		load_extension(hLibrary, TRUE, remote, NULL, extensionCommands);
-		stagelessExtensions = (MetsrvExtension*)((LPBYTE)stagelessExtensions->dll + stagelessExtensions->size);
+		dprintf("1");
```
Comment (max3raza) [inline c/meterpreter/source/server/server_setup_win.c]: This block should be reverted back (debug was added and bad indentation).

File: c/meterpreter/source/server/win/server_transport_tcp.c [hunk 5]
```diff
@@ -425,7 +425,6 @@ static DWORD packet_receive(Remote *remote, Packet **packet)
 		memcpy_s(packetBuffer, sizeof(PacketHeader), (LPBYTE)&encodedHeader, sizeof(PacketHeader));
 
 		LPBYTE payload = packetBuffer + sizeof(PacketHeader);
-
```
Comment (max3raza) [inline c/meterpreter/source/server/win/server_transport_tcp.c]: This file should be untouched.

File: c/meterpreter/source/common/config.h (hunk 1 above)
Comment (busterb) [reply c/meterpreter/source/common/config.h]: This PR is already pointing to the defcon-russia:master_update branch, so feel free to push these changes to that branch, and it will update the PR here.

Comment (max3raza) [reply c/meterpreter/source/common/config.h]: Done.

**Comment by bwatters-r7** [comment]
Where are we on this? Is it ready for testing, and/or where be the dragons?

**Comment by sempervictus** [comment]
Held up by me entirely - the portage from Python->Ruby/Msf is not going as quickly as i'd like. Q4 was a gauntlet and the CPU vendors are "just ruining my year." Trout slaps should be directed this way.

I should also push my deltas here, huh? I had to fix up some stuff for my fork, not sure if it applies here (i have bits of OJs code floating around as well as some of mine which affected the "socket" handle code IIRC).

**Comment by bwatters-r7** [comment]
Needs to include new patch send through covert channels?  Fix the dif?
defconrussiadns.diff ?

**Comment by sempervictus** [comment]
Got stalled, and a bit sidetracked with sniffer, will finish this up shortly.

**Comment by bwatters-r7** [comment]
There's a second portion to this.  Look at your email.

**Comment by akefallonitis** [comment]
Any update on this one ? DNS could very useful why its stalled ?

**Comment by sempervictus** [comment]
Because most of us work on this in our spare time, and that spare time 
comes in small increments and at long intervals. Be glad to trade some of 
this compliance work im doing for time to code. :)

-B

**Comment by akefallonitis** [comment]
Of course mate i just asked about the status cause its stalled for almost a year now. Thanks for your answer

**Comment by busterb** [comment]
I think this should start fresh once some of the newer work happening in the last few weeks is landed. Going to close this for now.

**Comment by adfoster-r7** [comment]
Does anyone have any details on the latest here? Were there any additional PRs put up anywhere to cross-reference? i.e. anything on the Metasploit-framework side of things?

Just wanting to see how much remaining work there was to get this over the line 👀 

Maybe @bwatters-r7 you have the latest? 😄

**Comment by sempervictus** [comment]
I have a branch with this stuff, and another one with the DNSCat version - both are quite old however, years of merge-forward if we were to use them. The stagers involved here are non-trivial and probably more difficult for me to reimplement than the actual transport... reason i say "reimplement" is that this code originated from russia and i dont know whether there are implications for merging that into FOSS much less commercial products in the West given GH freezing russian accounts and so forth.

However, now that we have native DNS services upstream, the generic effort should be a lot easier than it was back then :)
